anxcloud/resource_dns_record.go

```diff
@@ -0,0 +1,225 @@
+package anxcloud
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"go.anx.io/go-anxcloud/pkg/api"
+	"go.anx.io/go-anxcloud/pkg/api/types"
+	clouddnsv1 "go.anx.io/go-anxcloud/pkg/apis/clouddns/v1"
+	"go.anx.io/go-anxcloud/pkg/utils/object/compare"
+)
+
+var syncDNSRecordOps sync.Mutex
+var dnsRecordSkipMutexLock = providerContextKey("dns-record-skip-mutex-lock")
+
+func resourceDNSRecord() *schema.Resource {
+	return &schema.Resource{
+		Description:   "This resource allows you to create DNS records for a specified zone. Records of type TXT might be unstable for now.",
+		CreateContext: resourceDNSRecordCreate,
+		ReadContext:   resourceDNSRecordRead,
+		UpdateContext: resourceDNSRecordUpdate,
+		DeleteContext: resourceDNSRecordDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(time.Minute),
+			Read:   schema.DefaultTimeout(time.Minute),
+			Update: schema.DefaultTimeout(time.Minute),
+			Delete: schema.DefaultTimeout(time.Minute),
+		},
+		Schema: schemaDNSRecord(),
+	}
+}
+
+func resourceDNSRecordCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	defer lockRecordContextIfNeeded(ctx)()
+
+	a := m.(providerContext).api
+
+	r := dnsRecordFromResourceData(d)
+
+	// try to import
+	if ret, err := findDNSRecord(ctx, a, r); api.IgnoreNotFound(err) != nil {
+		return diag.FromErr(err)
+	} else if err == nil {
+		// DNS Record found
+		d.SetId(ret.Identifier)
+		return resourceDNSRecordRead(ctx, d, m)
+	}
+
+	// not found -> create new zone
+
+	if err := a.Create(ctx, &r); err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(r.Identifier)
+
+	return resourceDNSRecordRead(ctx, d, m)
+}
+
+func resourceDNSRecordRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags []diag.Diagnostic
+
+	a := m.(providerContext).api
+
+	r := dnsRecordFromResourceData(d)
+	r, err := findDNSRecord(ctx, a, r)
+
+	if api.IgnoreNotFound(err) != nil {
+		return diag.FromErr(err)
+	} else if err != nil {
+		d.SetId("")
+		return diags
+	}
+
+	if err := d.Set("type", r.Type); err != nil {
+		diags = append(diags, diag.FromErr(err)...)
+	}
+	if err := d.Set("name", r.Name); err != nil {
+		diags = append(diags, diag.FromErr(err)...)
+	}
+	if err := d.Set("zone_name", r.ZoneName); err != nil {
+		diags = append(diags, diag.FromErr(err)...)
+	}
+	if err := d.Set("region", r.Region); err != nil {
+		diags = append(diags, diag.FromErr(err)...)
+	}
+	// remove quotes from txt rdata to prevent double, tripple, ... quoted data
+	rData := r.RData
```

Comment:
move this rData handling above the `d.Set` blocks, it's an alien thing in a group of `d.Set`s and easily gets lost while reading the code